compiler/rustc_mir/src/interpret/terminator.rs

```diff
@@ -230,37 +256,38 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
             }
         };
 
+        let get_abi = |this: &Self, instance_ty: Ty<'tcx>| match instance_ty.kind() {
+            ty::FnDef(..) => instance_ty.fn_sig(*this.tcx).abi(),
+            ty::Closure(..) => Abi::RustCall,
+            ty::Generator(..) => Abi::Rust,
+            _ => span_bug!(this.cur_span(), "unexpected callee ty: {:?}", instance_ty),
+        };
+
         // ABI check
-        let check_abi = |this: &Self, instance_ty: Ty<'tcx>| -> InterpResult<'tcx> {
-            if M::enforce_abi(this) {
-                let callee_abi = match instance_ty.kind() {
-                    ty::FnDef(..) => instance_ty.fn_sig(*this.tcx).abi(),
-                    ty::Closure(..) => Abi::RustCall,
-                    ty::Generator(..) => Abi::Rust,
-                    _ => span_bug!(this.cur_span(), "unexpected callee ty: {:?}", instance_ty),
-                };
-                let normalize_abi = |abi| match abi {
-                    Abi::Rust | Abi::RustCall | Abi::RustIntrinsic | Abi::PlatformIntrinsic =>
-                    // These are all the same ABI, really.
-                    {
-                        Abi::Rust
-                    }
-                    abi => abi,
-                };
-                if normalize_abi(caller_abi) != normalize_abi(callee_abi) {
-                    throw_ub_format!(
-                        "calling a function with ABI {} using caller ABI {}",
-                        callee_abi.name(),
-                        caller_abi.name()
-                    )
+        let check_abi = |callee_abi: Abi| -> InterpResult<'tcx> {
+            let normalize_abi = |abi| match abi {
+                Abi::Rust | Abi::RustCall | Abi::RustIntrinsic | Abi::PlatformIntrinsic =>
+                // These are all the same ABI, really.
+                {
+                    Abi::Rust
                 }
+                abi => abi,
+            };
+            if normalize_abi(caller_abi) != normalize_abi(callee_abi) {
+                throw_ub_format!(
+                    "calling a function with ABI {} using caller ABI {}",
+                    callee_abi.name(),
+                    caller_abi.name()
+                )
             }
             Ok(())
         };
 
         match instance.def {
             ty::InstanceDef::Intrinsic(..) => {
-                check_abi(self, instance.ty(*self.tcx, self.param_env))?;
+                if M::enforce_abi(self) {
```

Comment:
Why did you move the `enforce_abi` check out of the `check_abi` closure?

Reply:
That can avoid calling `get_abi()`.

Reply:
Ah, so it's a (small) perf thing? Sure.

Thinking more about this... we still do the "unwinding is UB when caller or callee don't support it" check. So just disabling the ABI check will not be enough to support the testcases this is meant for, will it?

Reply:
> Thinking more about this... we still do the "unwinding is UB when caller or callee don't support it" check. So just disabling the ABI check will not be enough to support the testcases this is meant for, will it?

I think it's enough: https://github.com/rust-lang/miri/commit/cd8f2a57ccadd2d2348d2ae39ed6d577e8cc18d8
Although I have to remove some tests related to function pointers, or that I'm not sure how to *rewrite* without function pointers, since they are always allowed to unwind, in `fn_ptr_codegen_fn_attr_flags()`.

Reply:
> Although I have to remove some tests related to function pointers, or that I'm not sure how to rewrite without function pointers, since function pointers are always allowed to unwind, in fn_ptr_codegen_fn_attr_flags().

That seems like a bug; a call to an `extern "C" fn()` should not be allowed to unwind.